### ISSUE TITLE
JsonParsing using Object binding is extremely slow, unless parsed into a JsonObject first.

### ISSUE THREAD

**Issue by galli-leo** [cbeust/klaxon#154]
I wanted to use this library, so I didn't have to implement parsing a dictionary into a kotlin object myself.

Unfortunately, the library is extremely slow when using object binding. The following code, takes several second to parse the json:

```
val string = """{"message":"This is a test message!"}"""
val errorMessage = Klaxon().parse<ErrorMessageData>(string)
```

However, if I first parse the string into a jsonObject and then parse from there, it works really fast:

```
val string = """{"message":"This is a test message!"}"""
val obj = Klaxon().parseJsonObject(StringReader(string))
val errorMessage = Klaxon().parseFromJsonObject<ErrorMessageData>(obj)
```

That doesn't make much sense to me.

Testing was done on a Google Glass 2 (Android Kitkat 4.4) with the latest Kotlin runtime.

**Comment by cbeust** [comment]
Hi,

Interesting, I'll see if I can replicate these results and I'll get back to you.

**Comment by galli-leo** [comment]
It seems like the issue is due to threading? I am now getting bad performance even with the JsonObject. (Up to 13 seconds for parsing the above snippet).

**Comment by cbeust** [comment]
What do you mean by threading? Klaxon is single threaded.

**Comment by cbeust** [comment]
Something is indeed wrong, this:

```
        class ErrorMessageData(val message: String)
        val string = """{"message":"This is a test message!"}"""
        val t = measureTimeMillis {
            val errorMessage = Klaxon().parse<ErrorMessageData>(string)
        }
        println("Time: $t ms")
```

Takes 1.2 seconds in average for me. Definitely not right.

**Comment by galli-leo** [comment]
Sorry, I should have been clearer. The code from where I am calling is on a separate thread.

Could the issue be, the properties in the "constructor"? (Not sure what you call them, pretty new to kotlin).

**Comment by plannigan** [comment]
I looked into how the two ways are different.

* `parse<ErrorMessageData>()` uses reflection to see if `ErrorMessageData` has any attributes with the `@Json` annotation before calling `parseFromJsonObject<ErrorMessageData>()`.

*`parseJsonObject()` doesn't have any type information about the target type, so the reflection step is skipped.

I ran the code @cbeust posted. I saw 1.2 seconds once, but after that was getting between .7 and .9 seconds. Changing parse<T>() from having `parser(T::class)` to `parser()` drops the time to under .5 seconds. Which is effectively the same as what I get for the method that parses to JsonObject first.

In this case, there are not annotations, so the initial reflection step is wasted time and reflection is always slow. I haven't spent enough time with the code to know how this could be optimized for the case when the target class doesn't use annotations. However, a simple solution could be to add a boolean argument that would let the user specify if annotations are used.

**Comment by gavinwilliams** [comment]
I'm having the same issue with both serialising and deserialising to and from JSON using simple hashmaps. 

The first call is slow (about 2 - 3 seconds), subsequent calls are faster.

**Comment by spapas** [comment]
Hello, any news on this issue?

**Comment by Kerooker** [comment]
Any plans on improving these performance issues?

**Comment by rchampa** [comment]
Still happens on 5.0.12

**Comment by cbeust** [comment]
Initial investigations seem to point to a problem in `KClass::memberProperties`.

```
        class ErrorMessageData(val message: String)
        val r = measureTimeMillis {
            ErrorMessageData::class.memberProperties
        }
        println("Time: $r ms")
```

Produces:

```
Time: 1126 ms
```

I'm going to file a bug for the Kotlin team and see what they have to say.

**Comment by cbeust** [comment]
Done: https://youtrack.jetbrains.com/issue/KT-33760

**Comment by Kerooker** [comment]
In the meantime, perhaps we should go for the java alternative?

**Comment by cbeust** [comment]
I started looking into this and it's not going to be easy: Klaxon uses `kotlin-reflect` pervasively so that change has a big impact on the code base.

Not saying I won't do it, just that I regret not having used a layer of indirection for reflection :-)

**Comment by Kerooker** [comment]
Perhaps you can change that bit to java and use `toKFunction` or something
like that?

**Comment by cbeust** [comment]
The problem is that all these types from `kotlin-reflect` cannot be constructed from outside `kotlin-reflect`, so I'm going to need to create a parallel hierarchy and transfer objects from Java reflection to that hierarchy...

**Comment by cbeust** [comment]
I spent some time working on this, trying to rip out as much of `kotlin-reflect` as I could and I'm hitting a dead end where using Java reflection makes me lose some type information (because of erasure).

Will keep digging but I think we're going to have to wait for a fix from the Kotlin team.

**Comment by cbeust** [comment]
Alexander, from the Kotlin team, commented:  

> The initialization of kotlin-reflect indeed takes a lot of time, but subsequent accesses should be a lot faster. Unfortunately, this can't be fixed quickly, a major refactoring of the whole kotlin-reflect implementation is required. Meanwhile, you might be interested in kotlinx.reflect.lite which has a different API from kotlin-reflect, but implementation is much faster (there's a pull request in progress where a lot more features are added, you can try building it locally and see if anything from Klaxon's use cases is missing).

I'll take a look at `kotlinx.reflect.lite`.

**Comment by TitanKing** [comment]
I do recommend using https://github.com/square/moshi for now. Tested it and works much faster.

**Comment by alisonmoura** [comment]
I'm facing the same issue. **@cebeust** how is it going with `kotlinx.reflect.lite`? Hope that this issue get solved, it's a pretty good lib.

**Comment by Displee** [comment]
Same issue.

**Comment by fabiosussetto** [comment]
Can this caveat be added to the project readme? I was lucky to read this thread, I was going to use this in a production application before reading about this serious issue, which is really a deal breaker. I appreciate it's not your "fault" (if it can be even consider that), so thanks for your hard work on this!

**Comment by alisonmoura** [comment]
I agree with @fabiosussetto. In my case I had to rewrite the feature, cause I found the thread after looking for a solution for this issue.
